### ISSUE TITLE
Some sort of NamedArray problem?

### ISSUE THREAD

**Issue by luciansmith** [sys-bio/roadrunner#873]
When testing 2.1.0 with tellurium 2.2.1, I ran into a problem with the attached python script from Joe from the modeling workshop
[parameter_fitting_exercises_and_solutions.py.txt](https://github.com/sys-bio/roadrunner/files/7307442/parameter_fitting_exercises_and_solutions.py.txt)
 
The problem occurs on line 165.  Here's lines 164-5 for a bit of context:

```
    ax = fig.add_subplot(spec[irow, icol])
    ax.scatter(times, residualsArr[:, idx+1])
```

Originally, I got an error talking about multiple libraries being loaded from numpy, so I uninstalled numpy completely and reinstalled it.  After that, I got this error:

"Error: 1599: Unexpected type passed to NamedArrayObject_Finalize for the args parameter"

Given that @CiaranWelsh has been doing a lot of work to upgrade NamedArray, perhaps this is the problem?

I can say that when debugging the above, I can successfully get a NamedArray out of residualsArr[:,idx+1], so that's not the problem; presumably the problem is trying to convert that into whatever ax.scatter needs.

If I add:

```
import numpy
residualsArr = numpy.array(residualsArr)
```
just before the loop, I no longer fail the same way (instead I'm getting a CVODE 'failed to converge' error, which is presumably a different problem entirely.)

**Comment by hsauro** [comment]
Sounds like this needs to be fixed as soon as possible.

**Comment by CiaranWelsh** [comment]
This is certainly related to my attempts to fix the issues described in #869. I'll work on this today and if its not sorted by the time you guys (@luciansmith @hsauro ) wake up then perhaps we should have a short call to talk about the problem.

**Comment by CiaranWelsh** [comment]
This comment grew with debugging, but I'll leave all the steps in so you can follow my train of thought. 

Instructions [here](https://numpy.org/doc/stable/user/basics.subclassing.html) describe the three types of operations that should be dealt with by the numpy hook `__array_finalize__`. Copied here for convenience: 

```
1. Explicit constructor call - as in MySubClass(params). This is the usual route to Python instance creation.
2. View casting - casting an existing ndarray as a given subclass
3. New from template - creating a new instance from a template instance. Examples include returning slices from a subclassed array, creating return types from ufuncs, and copying arrays. See Creating new from template for more details
```
Since the error occurs in `File "C:\Miniconda3\envs\py39\lib\site-packages\numpy\ma\core.py", line 3652,` which is 
```
return ndarray.view(self, self._baseclass)
```
I'd say we're looking at a problem with view casting the NamedArray. The class is a MaskedArray so I was able to reproduce the problem with a call to `MaskedArray.view(NamedArray)`.

**Comment by CiaranWelsh** [comment]
Given the three points above, I previously modified [NamedArray_Finalize](https://github.com/sys-bio/roadrunner/blob/develop/wrappers/Python/roadrunner/PyUtils.cpp#L1570) to have three digestible parts, rather than one large function. The problem here, is that I haven't accounted for types other than `NamedArray` or `PyArray` in view casting. So a quickfix is to mimic old behaviour and put other subclasses of PyArray down the same path as Py_NONE. Importantly, I suspect this will have other problems associated with it, but these problems would have been present before release 2.1.0.

**Comment by CiaranWelsh** [comment]
Fixed in #884
